### ISSUE TITLE
error: failed to load language plugin python: python plugin [c:\pulumi\bin\pulumi-language-python.exe] did not begin responding to RPC connections

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#1934]
A user is seeing a failure to locate the python language plugin when using Windows + Python + `--local`.

```
C:\SDS Work\Workspaces\SDS_Services\Main\Build\Pulumi>pip install -r ./requirements.txt
Requirement already satisfied: pulumi>=0.15.0 in c:\program files (x86)\python3.6.5\lib\site-packages (from -r ./requirements.txt (line 1)) (0.15.0)
Requirement already satisfied: pulumi_aws>=0.15.0 in c:\program files (x86)\python3.6.5\lib\site-packages (from -r ./requirements.txt (line 2)) (0.15.0)
Requirement already satisfied: grpcio>=1.9.1 in c:\program files (x86)\python3.6.5\lib\site-packages (from pulumi>=0.15.0->-r ./requirements.txt (line 1)) (1.14.2)
Requirement already satisfied: protobuf>=3.6.0 in c:\program files (x86)\python3.6.5\lib\site-packages (from pulumi>=0.15.0->-r ./requirements.txt (line 1)) (3.6.1)
Requirement already satisfied: six>=1.11.0 in c:\program files (x86)\python3.6.5\lib\site-packages (from pulumi>=0.15.0->-r ./requirements.txt (line 1)) (1.11.0)
Requirement already satisfied: setuptools in c:\program files (x86)\python3.6.5\lib\site-packages (from protobuf>=3.6.0->pulumi>=0.15.0->-r ./requirements.txt (line 1)) (39.0.1)

C:\SDS Work\Workspaces\SDS_Services\Main\Build\Pulumi>
C:\SDS Work\Workspaces\SDS_Services\Main\Build\Pulumi>pulumi stack init dev1
Created stack 'dev1'.

C:\SDS Work\Workspaces\SDS_Services\Main\Build\Pulumi>pulumi config set aws:region us-east-1

C:\SDS Work\Workspaces\SDS_Services\Main\Build\Pulumi>pulumi up
Previewing update of stack 'dev1'
error: failed to load language plugin python: python plugin [c:\pulumi\bin\pulumi-language-python.exe] did not begin responding to RPC connections

C:\SDS Work\Workspaces\SDS_Services\Main\Build\Pulumi>
```

**Comment by lukehoban** [comment]
@swgillespie Could you add details of the investigation you did on trying to reproduce this?

**Comment by swgillespie** [comment]
I tried to repro this on a Win 10 Home machine yesterday and was not successful. Here are the configurations that I tried:

1. Pulumi 0.15.2, Python 2.7, deploying the `aws-python` template. Worked fine.
2. Pulumi 0.15.2, Python 3.7, deploying the `aws-python` template. Did not work fine, but for known reasons (https://github.com/pulumi/pulumi-terraform/issues/184).

Next thing to do is to inspect to the code to see why the language host might fail to start an RPC server.

**Comment by swgillespie** [comment]
Here's exactly my procedure for each configuration, starting in an empty directory

1. `pulumi new aws-python`
2. `pulumi stack init aws-win`
3. `pulumi config set aws:region us-west-1`
4. `pulumi update`

Each time I'm able to launch the language host successfully, which is where we are going off the rails in the bug.

If there is a consistent repro for this, would it be possible to capture the system calls using [ProcMon](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon)? That would be super helpful in understanding why we're failing to set up a gRPC session.

Looking at our own code, we are doing some [nontrivial logic](https://github.com/pulumi/pulumi/blob/c01023737e0b615f94527061477813fb13bdc9f8/pkg/resource/plugin/plugin.go#L184-L220) when determining whether or not our gRPC server is active. There could be something amiss there on Windows.

**Comment by pnandigam** [comment]
I'm seeing the issue using python 2.7.15 as well

**Comment by swgillespie** [comment]
@pnandigam OK, that's interesting. Do you mind posting the contents of the folder that you're running `pulumi update` from?

**Comment by pnandigam** [comment]
[Pulumi.zip](https://github.com/pulumi/pulumi/files/2384988/Pulumi.zip)

**Comment by swgillespie** [comment]
thank you! I'll keep at the investigation.

**Comment by swgillespie** [comment]
I'm still having trouble reproducing this issue - please let me know if it persists.

**Comment by lukehoban** [comment]
I'm going to close this since we haven't been able to repro and also recently made major changes to the Python Pulumi support.  @pnandigam if you are still having problems here - please do reach out again and we'll see if we can do a deeper dive on this.

**Comment by swgillespie** [comment]
@gitbalaji if you are using linux, can you post the results of running `strace -f pulumi up`? I can't repro this issue and I'm not sure how this happened.

**Comment by pgavlin** [comment]
I think that this may be proxy-related, and should be fixed with 0.16.6.

**Comment by pgavlin** [comment]
(see https://github.com/pulumi/pulumi/issues/2134 for details)

**Comment by lukehoban** [comment]
Closing out, since we believe this is fixed.  If anyone is still seeing this with `v0.16.6` or later, please leave a note with details and we will reopen.

**Comment by gitbalaji** [comment]
Thanks @lukehoban  it works after i installed 0.16.6 version

**Comment by lukehoban** [comment]
Great to hear - thanks for validating!

**Comment by carterwilliamson** [comment]
I started to see this yesterday. Initially I was running the python sdk on v0.17.27 but rolled back to .26 which did not seem to resolve the problem.
```
ERROR: logging before flag.Parse: I0730 07:46:28.700694   54847 log.go:56] Error closing python language plugin during shutdown; ignoring: 1 error occurred:
[environment-basic]             * operation not permitted
```
```bash
pulumi version
v0.17.26
```

```bash
pipenv run pip show pulumi
Name: pulumi
Version: 0.17.26
```

**Comment by lukehoban** [comment]
> ERROR: logging before flag.Parse: 

This is https://github.com/pulumi/pulumi/issues/2989 which should have been fixed in 0.17.27. See https://github.com/pulumi/pulumi/blob/master/CHANGELOG.md#01727-2019-07-29.

**Comment by carterwilliamson** [comment]
Thanks @lukehoban I'll test with that latest package and report back.
